### PR TITLE
Introducing Changeset workflow

### PR DESCRIPTION
**Pull request by taras**
## Hey, I just made a Pull Request!

This PR is a follow up to #1839. 

## TODO
- [x] Add @changesets/cli as a dependency in workspace root to allow using `yarn changeset`
- [x] Make all packages [linked](https://github.com/atlassian/changesets/blob/master/docs/linked-packages.md) so they'll be versioned in lockstep fashion
- [x] Add GitHub Action workflow to create release PR when changesets are present in master
- [x] Add documentation about creating changesets - https://github.com/taras/backstage/blob/tm/add-changesets/CONTRIBUTING.md#creating-changesets
- [ ] Add Changesets apps to the repository so it'll post a comment with summary of impact of change

## Questions:

1. Do we want to prevent PRs from being merged if they do not have a changeset? (if yes, I need to add another GitHub action workflow)
2. Where should we add documentation?

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

**Review by Rugvip — COMMENTED** [review]
Thanks! Looks promising :>

Regarding (1), I'm not sure tbh. Gut feeling is that we start out without enforcing until we fully understand what kind of changes we want to track and what the overhead is.

For (2) I think it makes sense to add a top-level entry in https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md

**Comment by taras** [comment]
I created an example branch to show what will happen when you make a [minor bump](https://github.com/taras/backstage/commit/803d1d497d8fbf17ce1908ca5651724ac51112db) in one of the linked packages and ran `changeset version` using the above changeset. I pushed a [commit](https://github.com/taras/backstage/commit/c43a2d2d636fcd101b1c3e3298f7142a4f8050c1) with all generated changes. Here is what the release PR will look like - you can see it as a [branch compare](https://github.com/taras/backstage/commit/c43a2d2d636fcd101b1c3e3298f7142a4f8050c1) with these changes combined.

File: .changeset/config.json
Comment (taras) [reply .changeset/config.json:6]: Maybe a better workaround for this would be to propose an addition to Changesets that makes lockstep versioning easier. We could add `lockstepped` flag to config which would automatically use monorepo packages as "linked" packages. 

This way we don't need to do anything funky. What do you think?

File: .github/workflows/changeset.yml
Comment (taras) [reply .github/workflows/changeset.yml:17]: There are no tags in the repo. I could use a commit's SHA.

**Comment by Rugvip** [comment]
@OrkoHunter, any idea what's up with the codecov report? This RP shouldn't be changing test coverage much

**Comment by OrkoHunter** [comment]
@Rugvip Agreed. Couple of things coming to mind -
* We should set the threshold to 1% at least. Changes less than 1% don't matter most of the times and shouldn't be reported.
* This branch is 36 commits behind master. And so when Codecov is comparing it with master, it's looking at those changes. - However, this should never be a problem, as it should only look at the files modified in this PR. I will take a look. 🙇

File: .changeset/config.json
Comment (Rugvip) [reply .changeset/config.json:6]: Yeah that'd definitely be a nicer solution!

**Review by Rugvip — APPROVED** [review]
@taras Thank you for setting this up! :grin:

If I understand correctly, the only impact of merging this would be that we'll get a PR opened by the bot to make a new release, and then we can hit merge on that at any point? I'm also assuming that that release will be a non-prerelease, probably 0.1.1? And also that it's a bit of an effort to set up prereleases with Changesets.

If that's the case I proposed that we merge this and make sure the PR looks alright, but that we don't use that PR for making a release just yet. Reason being that we'd want to look over the versioning in the team, and figure out if we want to do non-prerelease just yet. We're definitely not gonna try to wait for some 1.0-stable release, but I think we can use the first non-prerelease as a marker that we're making the project more stable, and as part of that we'd want to go through and make sure we're happy with some of the existing surfaces. Perhaps the main piece being that the routing API and interaction between plugins and the app is still in a bit of flux.

**Comment by taras** [comment]
> If I understand correctly, the only impact of merging this would be that we'll get a PR opened by the bot to make a new release, and then we can hit merge on that at any point?

Yes

> I'm also assuming that that release will be a non-prerelease, probably 0.1.1? And also that it's a bit of an effort to set up prereleases with Changesets.

It's just that prepreleases are complicated. They have a disclaimer to that effect in [their docs](https://github.com/atlassian/changesets/blob/master/docs/prereleases.md): `Warning! Prereleases are very complicated! Using them requires a thorough understanding of all parts of npm publishes. Mistakes can lead to repository and publish states that are very hard to fix.`.

> If that's the case I proposed that we merge this and make sure the PR looks alright, but that we don't use that PR for making a release just yet.

I would suggest that we install the bot after we merge this PR because it makes it easier to understand how impact of the specific change will impact versions. You don't need to merge that PR, it just stays there in case you're ready to merge something. Like this https://github.com/thefrontside/bigtest/pull/535

File: .changeset/config.json
Comment (taras) [reply .changeset/config.json:6]: I'm not the only person who had this idea https://github.com/atlassian/changesets/issues/364

**Comment by taras** [comment]
@Rugvip I added instructions for contributors.

File: CONTRIBUTING.md
Comment (OrkoHunter) [inline CONTRIBUTING.md:77]:
```suggestion
We use [changesets](https://github.com/atlassian/changesets) to help us prepare releases. It helps us make sure that every package affected by a change gets a proper version number and an entry in its `CHANGELOG.md`. To make the process of generating releases easy. it helps when contributors include changesets with their pull requests.
```

**Comment by Andarist** [comment]
Hi! Changesets maintainer here - just wanted to comment on prereleases. They work great and are not that complicated on their own - you can only bump into some obstacles if you start mixing/merging non-prerelease branches with a prerelease one.

Basically:
- u enter the prerelease mode
- from now on *all* new releases will be prereleases with the highest created bump type so far (so u can't have a non-prerelease package depending on a prerelease of another)
- when merging versioning PRs the used changesets are **added** to the list in `pre.json` and files **are kept** within the repo (whereas when using non-prerelease flow they are removed right away), you can check how this looks like in practice [here](https://github.com/emotion-js/emotion/tree/d949a37b3d5bc14af8147af0564c37d565f6c6ad/.changeset)
- you can later exit the prerelease mode which takes all aggregated changeset files up to this point and creates a final release will all information contained in them

This flow allows us to "review" (edit/remove) changesets during the prerelease mode at any time. For example:
- I introduce a new feature in the prerelease mode
- later I find a bug in it and I fix it (so I create patch changeset for it)
- later I can remove that patch changeset because it isn't actually interesting for the final release itself (after exiting the prerelease mode)

There are a couple of caveats with this:
- you should not merge your base branch to the prerelease one if it contains unreleased changesets 
- if you release a prerelease of a certain type (patch/minor/major) then you need to keep at least one changeset of such type, what I mean - weird things might happen if you release a major prerelease but later decide that you don't want that and you only want to have it as a minor

**Comment by Rugvip** [comment]
@Andarist Thank you for the info! I think we could perhaps merge this and then enter prerelease mode in a follow PR, just to have a bit of a separation of the process in the history.

**Comment by Rugvip** [comment]
Played around with this a bit, and as far as I can tell the package linking doesn't work with prereleases. Is that something that should work @Andarist?

Either way I'm thinking we can start off with merging this and then possibly enter prerelease mode and see what the release bot suggests.

**Comment by Andarist** [comment]
> Played around with this a bit, and as far as I can tell the package linking doesn't work with prereleases. Is that something that should work @Andarist?

It should work - we are using it in Emotion and I haven't found any particular problems with that. If you point me to commits on which you have observed things not working correctly I could take a look at it.

**Comment by taras** [comment]
@Rugvip you have to make sure your change forces an update. Otherwise, it might look like things are not working. Majority of dependencies start with ^ which covers patch releases. Try making an minor update and you'll see it bumping everything.

**Comment by Rugvip** [comment]
@taras Ah ok, that's gonna get a bit awkward though. Is there any way to force an update even for patch releases? Idea was to use minor versions while on `0.x.x` to represent breaking changes.

@Andarist I think the confusion around how version bumps work is why I thought it wasn't working.

**Comment by taras** [comment]
@Rugvip technically `^0.1.1-alpha.22` doesn't need to have version updated in dependencies to install `0.1.1-alpha.23`. This is the reason why versions are not updated - they're covered by the dependency declaration. 

image

If we want patch to increase everything across the board, then we should remove "^" from versions if we're expecting to be using prereleases for a while.

**Comment by Rugvip** [comment]
@taras sure, but I thought the point of linked packages was they were all gonna be updated together? ^_>

**Comment by Andarist** [comment]
> @taras sure, but I thought the point of linked packages was they were all gonna be updated together? ^_>

That would be "fixed" versioning. The concept of linked packages is slightly more complicated - it only ensures that packages in the set will get the highest version among the set **when** being released. The classification for the release itself doesn't change.

So instead of always bumping all packages in the set we try to minimize necessary releases.

I've been talking with Mitchell about some alternatives just last week (not for how linked works, but rather for how we sometimes bump dependant packages and how we sometimes dont) - if this doesn't satisfy your requirements then let's talk to try figure out some common ground.

**Comment by Andarist** [comment]
> If we want patch to increase everything across the board, then we should remove "^" from versions if we're expecting to be using prereleases for a while.

Right - the strategy to skip publishing when a dependency is still in range is OK for regular releases but we've found out that it't not so good for prereleases because they don't really have the same conceptual guarantee that `0.1.1-alpha.23` is a safe replacement for `0.1.1-alpha.22`. Prereleases are all about experimentation so we kinda think that anything beyond strict dependency ranges (dropping all modificators like `^`) lends itself to be unsafe for the end consumers. We are thinking how to best remedy this in Changesets.

**Comment by Rugvip** [comment]
Yeah I'm leaning towards exact deps, as in dropping the `^`. Makes more sense for the way we're treating the prereleases anyway.

**Comment by Andarist** [comment]
You'd only have to remember to bring back `^` when exiting the prerelease mode (assuming that it's what you want for regular releases). This is a bummer and easy to miss - therefore we were thinking about converting project to exact deps, but keeping info about specified range modificators so they could be brought back when exiting the prerelease mode. It's not super high on our priority list though, so I don't know if we implement this any soon. We were also thinking about slightly alternative way of doing prereleases - leveraging existing snapshots releases, just tailoring it a little bit to allow for somewhat regular release workflow. This would possibly solve many edge cases around existing prerelease mode by just removing a lot of complexity, statefulness etc. Either way - this also was just a random talk last week, but maybe we'll act on it some time in the future.

Bottom line is - versioning is hard, but the changesets concept is very solid (imho) 😅

**Comment by taras** [comment]
> versioning is hard, but the changesets concept is very solid (imho) 😅

I would agree. Since we started using Changesets for BigTest & Effection, we stopped worrying about releases. 

The only gap left to fill is creating pre-release distributions for Version Packages PRs. We punted on this, but it would be very helpful to auto publish the latest state of Version Packages PR branch.

**Comment by Andarist** [comment]
> The only gap left to fill is creating pre-release distributions for Version Packages PRs. We punted on this, but it would be very helpful to auto publish the latest state of Version Packages PR branch.

The mentioned snapshots releases could help you with that. You could release those on each commit there.

**Review by Rugvip — APPROVED** [review]
Time to :shipit: and then figure out the details. Agree that this seems like a robust strategy.

**Pull request merged**
